### PR TITLE
[Azure]:make our image MS certified

### PR DESCRIPTION
**Pull request by yaronkaikov**
Microsoft provides a certification test API that can be used to evaluate a given image's sanity. They also provide which test cases they run and expect us to comply with, as well as their recommendations for Azure Marketplace Images.

The following have been flagged within the VM self-test output for our
latest image:
```

"TestCaseName": "Swap Partition on OS Disk",
"Description": "Verifies that no Swap partitions are created on the OS disk.",
"Result": "Failed",
"ActualValue": "Swap space configured on the OS disk.",
```
Was fixed in #400

```
"TestCaseName": "Required Kernel Parameters",
"Description": "Verifies the following kernel parameters are set console=ttyS0, earlyprintk=ttyS0, rootdelay=300",
"Result": "Warning",
"ActualValue": "Missing Parameter: rootdelay=300\r\nMatched Parameter: console=ttyS0,earlyprintk=ttyS0",
---
"TestCaseName": "Client Alive Interval",
"Description": "It is recommended to set ClientAliveInterval to 180. On the application need, it can be set between 30 to 235. \nIf you are enabling the SSH for your end users this value must be set as explained.",
"Result": "Warning",
"ActualValue": "120",
```
fixed in this PR

Also fixing Azure image to be built based on Ubuntu:22.04 - 6b1b20b5c7d965da1d4d188a4a06a13c5035efee

Closes: scylladb/scylla-pkg#3172

### CODE REVIEW

**Review by fruch — APPROVED** [review]

**Comment by yaronkaikov** [comment]
@fruch Added to first commit additional change, please review. will merge after verification

**Comment by yaronkaikov** [comment]
@fruch It seems that due https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts/job/artifacts-azure-image-test/18/ started failing with:
```
04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >     ret = func(*args, **kwargs)
04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/session.pyx", line 222, in ssh2.session.Session.userauth_publickey_fromfile
04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/utils.pyx", line 166, in ssh2.utils.handle_error_codes
```
I guess we need a change also in SCT?

**Comment by fruch** [comment]
> @fruch It seems that due https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts/job/artifacts-azure-image-test/18/ started failing with:
> ```
> 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >     ret = func(*args, **kwargs)
> 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/session.pyx", line 222, in ssh2.session.Session.userauth_publickey_fromfile
> 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/utils.pyx", line 166, in ssh2.utils.handle_error_codes
> ```
> I guess we need a change also in SCT?

You didn't run this job before merging 6b1b20b5c7d965da1d4d188a4a06a13c5035efee ?

Could be that we need a change in SCT, we'll need to make sure we pass the hack needed into user-data (or other means), I think we are already doing that for AWS and GCP, will need to check why we are not doing so for Azure.

Meanwhile we might want to revert ? is that change already in 5.2 ?

**Comment by yaronkaikov** [comment]
> > @fruch It seems that due [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts/job/artifacts-azure-image-test/18/ started failing with:
> > ```
> > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >     ret = func(*args, **kwargs)
> > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/session.pyx", line 222, in ssh2.session.Session.userauth_publickey_fromfile
> > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/utils.pyx", line 166, in ssh2.utils.handle_error_codes
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I guess we need a change also in SCT?
> 
> You didn't run this job before merging [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) ?
> 
> Could be that we need a change in SCT, we'll need to make sure we pass the hack needed into user-data (or other means), I think we are already doing that for AWS and GCP, will need to check why we are not doing so for Azure.
> 
> Meanwhile we might want to revert ? is that change already in 5.2 ?

@fruch I didn't merge it :-) , I was verifying my changes before.  This change needs to backport to 5.2 as well, so let's try to fix before merging this one

**Comment by fruch** [comment]
> > > @fruch It seems that due [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts/job/artifacts-azure-image-test/18/ started failing with:
> > > ```
> > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >     ret = func(*args, **kwargs)
> > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/session.pyx", line 222, in ssh2.session.Session.userauth_publickey_fromfile
> > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/utils.pyx", line 166, in ssh2.utils.handle_error_codes
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I guess we need a change also in SCT?
> > 
> > 
> > You didn't run this job before merging [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) ?
> > Could be that we need a change in SCT, we'll need to make sure we pass the hack needed into user-data (or other means), I think we are already doing that for AWS and GCP, will need to check why we are not doing so for Azure.
> > Meanwhile we might want to revert ? is that change already in 5.2 ?
> 
> @fruch I didn't merge it :-) , I was verifying my changes before. This change needs to backport to 5.2 as well, so let's try to fix before merging this one

nevermind, I thought the change is in a different PR 

I've rerun the test and reset the ssh configuration via Azure Portal, and according to the logs, our user is locked somehow.
```
sshd[942]: User scyllaadm not allowed because account is locked
```

so it's not SCT issue, it's something total in SMI part

**Comment by yaronkaikov** [comment]
> > > > @fruch It seems that due [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts/job/artifacts-azure-image-test/18/ started failing with:
> > > > ```
> > > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >     ret = func(*args, **kwargs)
> > > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/session.pyx", line 222, in ssh2.session.Session.userauth_publickey_fromfile
> > > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/utils.pyx", line 166, in ssh2.utils.handle_error_codes
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I guess we need a change also in SCT?
> > > 
> > > 
> > > You didn't run this job before merging [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) ?
> > > Could be that we need a change in SCT, we'll need to make sure we pass the hack needed into user-data (or other means), I think we are already doing that for AWS and GCP, will need to check why we are not doing so for Azure.
> > > Meanwhile we might want to revert ? is that change already in 5.2 ?
> > 
> > 
> > @fruch I didn't merge it :-) , I was verifying my changes before. This change needs to backport to 5.2 as well, so let's try to fix before merging this one
> 
> nevermind, I thought the change is in a different PR
> 
> I've rerun the test and reset the ssh configuration via Azure Portal, and according to the logs, our user is locked somehow.
> 
> ```
> sshd[942]: User scyllaadm not allowed because account is locked
> ```
> 
> so it's not SCT issue, it's something total in SMI part

Looks like it was related to one of my latest changes. fixed it now
https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/azure-image/16/console

**Comment by yaronkaikov** [comment]
> > > > > @fruch It seems that due [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts/job/artifacts-azure-image-test/18/ started failing with:
> > > > > ```
> > > > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >     ret = func(*args, **kwargs)
> > > > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/session.pyx", line 222, in ssh2.session.Session.userauth_publickey_fromfile
> > > > > 04:10:56  < t:2023-01-23 02:10:55,581 f:remote_libssh_cmd_runner.py l:66   c:RemoteLibSSH2CmdRunner p:ERROR >   File "ssh2/utils.pyx", line 166, in ssh2.utils.handle_error_codes
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > I guess we need a change also in SCT?
> > > > 
> > > > 
> > > > You didn't run this job before merging [6b1b20b](https://github.com/scylladb/scylla-machine-image/commit/6b1b20b5c7d965da1d4d188a4a06a13c5035efee) ?
> > > > Could be that we need a change in SCT, we'll need to make sure we pass the hack needed into user-data (or other means), I think we are already doing that for AWS and GCP, will need to check why we are not doing so for Azure.
> > > > Meanwhile we might want to revert ? is that change already in 5.2 ?
> > > 
> > > 
> > > @fruch I didn't merge it :-) , I was verifying my changes before. This change needs to backport to 5.2 as well, so let's try to fix before merging this one
> > 
> > 
> > nevermind, I thought the change is in a different PR
> > I've rerun the test and reset the ssh configuration via Azure Portal, and according to the logs, our user is locked somehow.
> > ```
> > sshd[942]: User scyllaadm not allowed because account is locked
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so it's not SCT issue, it's something total in SMI part
> 
> Looks like it was related to one of my latest changes. fixed it now Running verification - https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/azure-image/16/console

@fruch Verification Passed !

File: packer/scylla_install_image
Comment (fruch) [reply packer/scylla_install_image:217]: i.e. `99_scylla_client_alive.conf`

Comment (yaronkaikov) [reply packer/scylla_install_image:217]: Because `/etc/ssh/sshd_config.d/50-cloudimg-settings.conf` already have this configuration `ClientAliveInterval 120` , how adding a new configuration file will help us? i don't think it will override it

Comment (yaronkaikov) [reply packer/scylla_install_image:217]: btw, this change is only related to `Azure`, and this is based on their recommendations , so anyway it will not break AWS or GCP

Comment (fruch) [reply packer/scylla_install_image:217]: > now I saw your other comment ;-) If `ClientAliveInterval` is not available anymore, shouldn't the new configuration file will break the next releases as well? If I am not mistaken, the command in L#216 will be replaced only for a specific value, so if this is not available anymore we will not do anything, right?

yes exactly what I mean,

changing configuration in `.d` folder is always in alphabetically order, hence adding bigger number at the prefix of the files should be working.

**Review by fruch — DISMISSED** [review]
Looks o.k., 
But I still that for portability, we should "sed" existing configuration files, not in our control.

File: packer/scylla_install_image
Comment (syuu1228) [inline packer/scylla_install_image:219]: Why it is copying exiting GRUB code defined at [line 164 - 169](https://github.com/scylladb/scylla-machine-image/blob/b5da657fa207d0c0069df993068a46531381a6cd/packer/scylla_install_image#L164)?
I think these codes should merge like swap setup (it has cloud specific variable `swap_opt`).

**Review by syuu1228 — CHANGES_REQUESTED** [review]

File: packer/scylla_install_image
Comment (yaronkaikov) [reply packer/scylla_install_image]: @fruch Changed, please review

File: packer/scylla_install_image
Comment (yaronkaikov) [reply packer/scylla_install_image]: Changed, please review @syuu1228

**Review by fruch — APPROVED** [review]
LGTM

**Review by syuu1228 — DISMISSED** [review]
LGTM

**Comment by yaronkaikov** [comment]
Verified:
https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/23/

**Comment by yaronkaikov** [comment]
With latest Azure image:
```
{"TrackingId":"1e12e62b-8bd3-4f3d-8ee4-34cba0cc2948","Response":"{\"SchemaVersion\":1,\"AppCertificationCategory\":\"Microsoft Single VM Certification\",\"ProviderID\":\"050DE427-2A99-46D4-817C-5354D3BF2AE8\",\"OSName\":\"Ubuntu 22.04\",\"OSDistro\":\"Ubuntu 22.04.1 LTS\",\"KernelVersion\":\"5.15.0-1031-azure\",\"KernelFullVersion\":\"Linux version 5.15.0-1031-azure (buildd@lcy02-amd64-010) (gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, GNU ld (GNU Binutils for Ubuntu) 2.38) #38-Ubuntu SMP Mon Jan 9 12:49:59 UTC 2023\\n\",\"OSVersion\":\"22.04\",\"CreatedDate\":\"01/30/2023 07:52:29\",\"TestResult\":\"Pass\",\"APIVersion\":\"1.7\",\"Tests\":[{\"TestID\":\"48\",\"TestCaseName\":\"Bash History\",\"Description\":\"Bash history files should be cleared before creating the VM image.\",\"Result\":\"Passed\",\"ActualValue\":\"1 bash history file(s) exists, each file size is upto 1KB.\",\"RequiredValue\":\"1024\"},{\"TestID\":\"39\",\"TestCaseName\":\"Linux Agent Version\",\"Description\":\"Azure Linux Agent Version 2.2.41 and above should be installed.\",\"Result\":\"Passed\",\"ActualValue\":\"2.9.0\",\"RequiredValue\":\"2.2.41\"},{\"TestID\":\"40\",\"TestCaseName\":\"Required Kernel Parameters\",\"Description\":\"Verifies the following kernel parameters are set console=ttyS0, earlyprintk=ttyS0, rootdelay=300\",\"Result\":\"Passed\",\"ActualValue\":\"Matched Parameter: console=ttyS0,earlyprintk=ttyS0,rootdelay=300\",\"RequiredValue\":\"console=ttyS0#earlyprintk=ttyS0#rootdelay=300\"},{\"TestID\":\"41\",\"TestCaseName\":\"Swap Partition on OS Disk\",\"Description\":\"Verifies that no Swap partitions are created on the OS disk.\",\"Result\":\"Passed\",\"ActualValue\":\"No. of Swap Partitions: 0\",\"RequiredValue\":\"swap\"},{\"TestID\":\"42\",\"TestCaseName\":\"Root Partition on OS Disk\",\"Description\":\"It is recommended that a single root partition is created for the OS disk.\",\"Result\":\"Passed\",\"ActualValue\":\"Root Partition: 1\",\"RequiredValue\":\"1\"},{\"TestID\":\"44\",\"TestCaseName\":\"OpenSSL Version\",\"Description\":\"OpenSSL Version should be >=0.9.8.\",\"Result\":\"Passed\",\"ActualValue\":\"3.0.2\",\"RequiredValue\":\"0.9.8\"},{\"TestID\":\"45\",\"TestCaseName\":\"Python Version\",\"Description\":\"Python version 2.6+ is highly recommended. \",\"Result\":\"Passed\",\"ActualValue\":\"3.10.6\",\"RequiredValue\":\"2.6\"},{\"TestID\":\"46\",\"TestCaseName\":\"Client Alive Interval\",\"Description\":\"It is recommended to set ClientAliveInterval to 180. On the application need, it can be set between 30 to 235. \\nIf you are enabling the SSH for your end users this value must be set as explained.\",\"Result\":\"Passed\",\"ActualValue\":\"180\",\"RequiredValue\":\"ClientAliveInterval 180\"},{\"TestID\":\"49\",\"TestCaseName\":\"OS Architecture\",\"Description\":\"Only 64-bit operating system should be supported.\",\"Result\":\"Passed\",\"ActualValue\":\"x86_64\\n\",\"RequiredValue\":\"x86_64,amd64\"},{\"TestID\":\"51\",\"TestCaseName\":\"Auto Update\",\"Description\":\"Identifies if Linux Agent Auto Update is enabled or not.\",\"Result\":\"Passed\",\"ActualValue\":\"# AutoUpdate.Enabled=y\\n\",\"RequiredValue\":\"Yes\"}]}"}
```

**Review by fruch — APPROVED** [review]
LGTM

**Review by syuu1228 — APPROVED** [review]
Review again for GRUB changes, LGTM

**Comment by yaronkaikov** [comment]
It seems like my patch breaks `next-machine-image`, strange since it was verified with https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/23/

Anyway, i dequeued this patch and will send a v2

**Comment by fruch** [comment]
> It seems like my patch breaks `next-machine-image`, strange since it was verified with https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/23/
> 
> Anyway, i dequeued this patch and will send a v2

could be a change in scylla-core, the failing script is from scylla-core

**Comment by yaronkaikov** [comment]
> > It seems like my patch breaks `next-machine-image`, strange since it was verified with https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/23/
> > Anyway, i dequeued this patch and will send a v2
> 
> could be a change in scylla-core, the failing script is from scylla-core

I ran https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/26/ from `scylla-machine-image` with `master` branch and it passed. The only new commits in `machine-image` were mine :-(

**Comment by fruch** [comment]
> > > It seems like my patch breaks `next-machine-image`, strange since it was verified with https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/23/
> > > Anyway, i dequeued this patch and will send a v2
> > 
> > 
> > could be a change in scylla-core, the failing script is from scylla-core
> 
> I ran https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/26/ from `scylla-machine-image` with `master` branch and it passed. The only new commits in `machine-image` were mine :-(

very strange indeed, since I don't see anything that would break it.

**Comment by yaronkaikov** [comment]
> > > > It seems like my patch breaks `next-machine-image`, strange since it was verified with https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/23/
> > > > Anyway, i dequeued this patch and will send a v2
> > > 
> > > 
> > > could be a change in scylla-core, the failing script is from scylla-core
> > 
> > 
> > I ran https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/next-machine-image/26/ from `scylla-machine-image` with `master` branch and it passed. The only new commits in `machine-image` were mine :-(
> 
> very strange indeed, since I don't see anything that would break it.

@fruch I Found the reason (I guess), Setting for AWS and GCP 
`
grub = re.sub(r'^GRUB_CMDLINE_LINUX="(.+)"$',
                  fr'GRUB_CMDLINE_LINUX="\1 net.ifnames=0 clocksource=tsc tsc=reliable {kernel_opt}"', grub,
                  flags=re.MULTILINE)
`
instead of `GRUB_CMDLINE_LINUX_DEFAULT` as it was before

**Pull request merged**
